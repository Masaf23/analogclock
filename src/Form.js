import React, { Component } from 'react';

class Form extends Component {

    constructor(props) {
        super(props);
        this.state = {
            options: this.props.defaultOptions
        };
    }

    setClockSize = (event) => {
        event.preventDefault();
        let width = event.target.value + 'px';
        this.setState({ options: { ...this.state.options, width } })
    }

    setCustomTime = (event) => {
        let useCustomTime = event.target.value === 'yes';
        this.setState({ options: { ...this.state.options, useCustomTime } })
    }


    setBorderReq = (event) => {
        let border = event.target.value === 'yes';
        this.setState({ options: { ...this.state.options, border } })
    }

    setColor = (event) => {
        event.preventDefault();
        this.setState({ options: { ...this.state.options, [event.target.name]: event.target.value } })
    }

    setHandColor = (event) => {
        event.preventDefault();
        let handColors = { ...this.state.options.handColors };
        handColors[event.target.name] = event.target.value;
        this.setState({ options: { ...this.state.options, handColors } });
    }

    buildClock = (event) => {
        event.preventDefault();
        event.stopPropagation();
        this.props.customizeClock(this.state.options);
    }

    randomClock = (event) => {
        event.preventDefault();
        event.stopPropagation();
        let baseColor = this.getRandomColor();
        let borderColor = this.getRandomColor();
        let centerColor = this.getRandomColor();
        let centerBorderColor = this.getRandomColor();
        let handColors = {
            second: this.getRandomColor(),
            minute: this.getRandomColor(),
            hour: this.getRandomColor()
        }
        this.setState({ options: { ...this.state.options, borderColor, baseColor, centerColor, centerBorderColor, handColors } }, () => this.buildClock(event));
    }

    getRandomColor() {
        var letters = '0123456789ABCDEF';
        var colorCode = '#';
        for (var i = 0; i < 6; i++) {
            colorCode += letters[Math.floor(Math.random() * 16)];
        }
        return colorCode;
    }
    
    render() {
        return (
            <form className="col-12 form">
                <div className="row col-12">
                    <h4><i>Clock Options</i></h4>
                </div>
                <div className="row">
                    <div className="col-4">
                        <div className="form-group">
                            <label htmlFor="clock-size">Clock size (px): {this.state.options.width}</label>
                            <input type="range" min="50" max="500" className="form-control-range" id="clock-size" onChange={this.setClockSize} />
                        </div>
                        <div className="form-group">
                            <label htmlFor="border-req-radios">Use Custom Time?</label>
                            <div className="form-check" name="custom-time-radios">
                                <input className="form-check-input" type="radio" name="custom-time-yes" id="custom-time-yes" value="yes" checked={this.state.options.useCustomTime === true} onChange={this.setCustomTime} />
                                <label className="form-check-label" htmlFor="custom-time-yes">Yes</label>
                            </div>
                            <div className="form-check">
                                <input className="form-check-input" type="radio" name="custom-time-no" id="custom-time-no" value="no" checked={this.state.options.useCustomTime === false} onChange={this.setCustomTime} />
                                <label className="form-check-label" htmlFor="custom-time-no">No</label>
                            </div>
                        </div>
                        <div className="form-group">
                            <label htmlFor="border-req-radios">Border Required?</label>
                            <div className="form-check" name="border-req-radios">
                                <input className="form-check-input" type="radio" name="border-req-yes" id="border-req-yes" value="yes" checked={this.state.options.border === true} onChange={this.setBorderReq} />
                                <label className="form-check-label" htmlFor="border-req-yes">Yes</label>
                            </div>
                            <div className="form-check">
                                <input className="form-check-input" type="radio" name="bored-req-no" id="bored-req-no" value="no" checked={this.state.options.border === false} onChange={this.setBorderReq} />
                                <label className="form-check-label" htmlFor="bored-req-no">No</label>
                            </div>
                        </div>
                    </div>
                    <div className="col-4">
                        <div className="form-group">
                            <label htmlFor="clock-border-color">Clock border color</label>
                            <input type="color"  name="borderColor" className="form-control" id="clock-border-color" maxLength="6" placeholder="2e2e2e" onChange={this.setColor} value={this.state.options.borderColor} />
                        </div>
                        <div className="form-group">
                            <label htmlFor="clock-base-color">Clock base color</label>
                            <input type="color" name="baseColor" className="form-control" id="clock-base-color" maxLength="6" placeholder="17a2b8" onChange={this.setColor}  value={this.state.options.baseColor}/>
                        </div>
                        <div className="form-group">
                            <label htmlFor="clock-center-color">Clock center color</label>
                            <input type="color" name="centerColor" className="form-control" id="clock-center-color" maxLength="6" placeholder="459cff" onChange={this.setColor} value={this.state.options.centerColor}/>
                        </div>
                        <div className="form-group">
                            <label htmlFor="clock-center-border-color">Clock center border color</label>
                            <input type="color" name="centerBorderColor" className="form-control" id="clock-center-border-color" maxLength="6" placeholder="fff" onChange={this.setColor} value={this.state.options.centerBorderColor}></input>
                        </div>
                    </div>
                    <div className="col-4">
                        <div className="form-group">
                            <label htmlFor="second-hand-color">Second Hand color</label>
                            <input type="color" name="second" className="form-control" id="second-hand-color" maxLength="6" placeholder="d81c7a" onChange={this.setHandColor} value={this.state.options.handColors.second}/>
                        </div>
                        <div className="form-group">
                            <label htmlFor="minute-hand-color">Minute Hand color</label>
                            <input type="color" name="minute" className="form-control" id="minute-hand-color" maxLength="6" placeholder="fff" onChange={this.setHandColor} value={this.state.options.handColors.minute}/>
                        </div>
                        <div className="form-group">
                            <label htmlFor="hour-hand-color">Hour Hand color</label>
                            <input type="color" name="hour" className="form-control" id="hour-hand-color" maxLength="6" placeholder="fff" onChange={this.setHandColor} value={this.state.options.handColors.hour}/>
                        </div>
                    </div>
                </div>
                <div className="row col-12 d-flex justify-content-center">
                    <button type="submit" className="btn btn-primary" onClick={this.buildClock}>Build Clock!</button>
                    &nbsp;
                <button type="submit" className="btn btn-warning" onClick={this.randomClock}>Surprise Me!</button>
                </div>
            </form>
        )
    }

}

export default Form;