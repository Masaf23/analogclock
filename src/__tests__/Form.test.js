import React from "react";
import Form from '../Form'
import { shallow } from "enzyme";

describe('Form component', () => {

    it('matches snapshot', () => {
        let options = {
            useCustomTime: false,
            width: "300px",
            border: true,
            borderColor: "#2e2e2e",
            baseColor: "#17a2b8",
            centerColor: "#459cff",
            centerBorderColor: "#fff",
            handColors: {
                second: "#d81c7a",
                minute: "#fff",
                hour: "#fff"
            },
        }
        const wrapper = shallow(<Form defaultOptions={options} />);
        expect(wrapper).toMatchSnapshot()
    });

})