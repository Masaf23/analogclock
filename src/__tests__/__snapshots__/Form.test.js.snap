// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component matches snapshot 1`] = `
<form
  className="col-12 form"
>
  <div
    className="row col-12"
  >
    <h4>
      <i>
        Clock Options
      </i>
    </h4>
  </div>
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="clock-size"
        >
          Clock size (px): 
          300px
        </label>
        <input
          className="form-control-range"
          id="clock-size"
          max="500"
          min="50"
          onChange={[Function]}
          type="range"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="border-req-radios"
        >
          Use Custom Time?
        </label>
        <div
          className="form-check"
          name="custom-time-radios"
        >
          <input
            checked={false}
            className="form-check-input"
            id="custom-time-yes"
            name="custom-time-yes"
            onChange={[Function]}
            type="radio"
            value="yes"
          />
          <label
            className="form-check-label"
            htmlFor="custom-time-yes"
          >
            Yes
          </label>
        </div>
        <div
          className="form-check"
        >
          <input
            checked={true}
            className="form-check-input"
            id="custom-time-no"
            name="custom-time-no"
            onChange={[Function]}
            type="radio"
            value="no"
          />
          <label
            className="form-check-label"
            htmlFor="custom-time-no"
          >
            No
          </label>
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="border-req-radios"
        >
          Border Required?
        </label>
        <div
          className="form-check"
          name="border-req-radios"
        >
          <input
            checked={true}
            className="form-check-input"
            id="border-req-yes"
            name="border-req-yes"
            onChange={[Function]}
            type="radio"
            value="yes"
          />
          <label
            className="form-check-label"
            htmlFor="border-req-yes"
          >
            Yes
          </label>
        </div>
        <div
          className="form-check"
        >
          <input
            checked={false}
            className="form-check-input"
            id="border-req-no"
            name="bored-req-no"
            onChange={[Function]}
            type="radio"
            value="no"
          />
          <label
            className="form-check-label"
            htmlFor="bored-req-no"
          >
            No
          </label>
        </div>
      </div>
    </div>
    <div
      className="col-4"
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="clock-border-color"
        >
          Clock border color
        </label>
        <input
          className="form-control"
          id="clock-border-color"
          maxLength="6"
          name="borderColor"
          onChange={[Function]}
          placeholder="2e2e2e"
          type="input"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="clock-base-color"
        >
          Clock base color
        </label>
        <input
          className="form-control"
          id="clock-base-color"
          maxLength="6"
          name="baseColor"
          onChange={[Function]}
          placeholder="17a2b8"
          type="input"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="clock-center-color"
        >
          Clock center color
        </label>
        <input
          className="form-control"
          id="clock-center-color"
          maxLength="6"
          name="centerColor"
          onChange={[Function]}
          placeholder="459cff"
          type="input"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="clock-center-border-color"
        >
          Clock center border color
        </label>
        <input
          className="form-control"
          id="clock-center-border-color"
          maxLength="6"
          name="centerBorderColor"
          onChange={[Function]}
          placeholder="fff"
          type="input"
        />
      </div>
    </div>
    <div
      className="col-4"
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="second-hand-color"
        >
          Second Hand color
        </label>
        <input
          className="form-control"
          id="second-hand-color"
          maxLength="6"
          name="second"
          onChange={[Function]}
          placeholder="d81c7a"
          type="input"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="minute-hand-color"
        >
          Minute Hand color
        </label>
        <input
          className="form-control"
          id="minute-hand-color"
          maxLength="6"
          name="minute"
          onChange={[Function]}
          placeholder="fff"
          type="input"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="hour-hand-color"
        >
          Hour Hand color
        </label>
        <input
          className="form-control"
          id="hour-hand-color"
          maxLength="6"
          name="hour"
          onChange={[Function]}
          placeholder="fff"
          type="input"
        />
      </div>
    </div>
  </div>
  <div
    className="row col-12 d-flex justify-content-center"
  >
    <button
      className="btn btn-primary"
      id="build"
      onClick={[Function]}
      type="submit"
    >
      Build Clock!
    </button>
    Â 
    <button
      className="btn btn-warning"
      id="surprise"
      onClick={[Function]}
      type="submit"
    >
      Surprise Me!
    </button>
  </div>
</form>
`;
